core/src/main/java/software/amazon/awssdk/retry/RetryPolicy.java

```diff
@@ -15,222 +15,106 @@
          */
-        boolean shouldRetry(AmazonWebServiceRequest originalRequest,
-                            AmazonClientException exception,
-                            int retriesAttempted);
-
-    }
+        Builder backoffStrategy(BackoffStrategy backoffStrategy);
 
-    /**
-     * The hook for providing custom back-off strategy to control the sleep time
-     * between retries.
-     */
-    public interface BackoffStrategy {
-        RetryPolicy.BackoffStrategy NO_DELAY = new BackoffStrategy() {
-            @Override
-            public long delayBeforeNextRetry(AmazonWebServiceRequest originalRequest,
-                                             AmazonClientException exception,
-                                             int retriesAttempted) {
-                return 0;
-            }
-        };
+        /**
+         * The backoff strategy configured with {@link #backoffStrategy()}.
+         */
+        BackoffStrategy backoffStrategy();
 
         /**
-         * Returns the delay (in milliseconds) before next retry attempt.
+         * Specifies the retry condition to use when retrying requests.
          *
-         * @param originalRequest
-         *            The original request object being executed. For
-         *            performance reason, this object is not a defensive copy,
-         *            and caller should not attempt to modify its data.
-         * @param exception
-         *            The exception from the failed request, represented as an
-         *            AmazonClientException object. There are two types of
-         *            exception that will be passed to this method:
-         *            <ul>
-         *              <li>AmazonServiceException (sub-class of
-         *                  AmazonClientException) indicating a service error
-         *              <li>AmazonClientException caused by an IOException when
-         *                  executing the HTTP request.
-         *            </ul>
-         *            Any other exceptions are regarded as unexpected failures
-         *            and are thrown immediately without any retry. For
-         *            performance reason, this object is not a defensive copy,
-         *            and caller should not attempt to modify its data.
-         * @param retriesAttempted
-         *            The number of times the current request has been attempted
-         *            (not including the next attempt after the delay).
+         * @see RetryCondition
+         * @see AndRetryCondition
+         * @see OrRetryCondition
+         * @see RetryOnErrorCodeCondition
+         * @see RetryOnStatusCodeCondition
+         * @see RetryOnExceptionsCondition
          *
-         * @return The delay (in milliseconds) before next retry attempt.
+         * @param retryCondition The retry condition
+         * @return This builder for method chaining
+         */
+        Builder retryCondition(RetryCondition retryCondition);
```

Comment:
You have maxNumberRetries and RetryCondition as separate things. I'd expect the maximum number of retries be part of the condition.

Reply:
It actually builds down to the RetryCondition but I do think it should be a separate field. Customers may want to just modify the number of retries without changing the other conditions that we retry on. This gives them an easy way to do that. If we make it required as part of the retry condition and force customers to use the AndRetryCondition its not really user friendly.

Reply:
I have gone back and forth on this myself. On the one hand it bothers me from an abstraction point of view but I totally get that it's common for customers to want to tweak the number of retries without messing too much with the other conditions on which it retries. I wonder if we can make this easier by returning a builder similiar to https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-opensdk/src/main/java/com/amazonaws/opensdk/retry/RetryPolicyBuilder.java with all the service defaults filled in and customers can tweak that as they need it. Perhaps we can expose a method on the builder that provides them the builder to modify.

Reply:
Matt and I talked more about this offline yesterday.

Basically numRetries is the maximum amount of retries per a given request. Its specified at the top level to allow for easy configuration as it is a value that is common to change. It also allows a easy way to specify an overall max for a given request.

Within each condition, you are given the number of attempts and additional context so you can still have a condition that does less retries. You would just do that within the shouldRetry method. 

We already kinda have a method for returning a builder with the defaults specified. If you just do RetryPolicy.builder().build() it will return a retry policy with all defaults and you then have a toBuilder().

Reply:
That doesn't work with service defaults though, ideally core should define no service specific default retry (outside of connection/io related failures) and all service retryable exceptions should be wired up by the service builder. Now we may not be able to get fully there by GA but our interfaces should be forward thinking with that in mind.

Reply:
We can sync offline on this further. I think we are on the same page.

Reply:
Sure thing. I'm free all day today.